### PR TITLE
Release management in CI

### PR DESCRIPTION
**Pull request by oguzkocer**
This PR adds the necessary configurations to handle the project's release management in CI. It closely follows the approach we've explored in [release-toolkit-demo project](https://github.com/wordpress-mobile/release-toolkit-demo) and discussed internally in our RFCs.

There are a few moving parts to be aware of:

1. There are new Buildkite triggers to start release management jobs. These are contained in `fastlane/lanes/release_management_in_ci.rb`. [We intend to remove these triggers and use `curl` instead.](https://github.com/wordpress-mobile/WordPress-Android/pull/18398/files#r1190589034)
2. There are new pipelines for each release management step, such as `.buildkite/code-freeze.yml`, `.buildkite/complete-code-freeze.yml` etc. These are pretty straightforward as they just call the relevant lane after some basic configuration.
3. There are a few new Buildkite helper commands. `.buildkite/commands/configure-git-for-release-management.sh` configures the git client, which we intend to move to our agents. `.buildkite/commands/checkout-release-branch.sh` checks out the release branch which is necessary because Buildkite starts the job in a detached state and we need to be on a branch to be able to commit and push the changes. Similarly `.buildkite/commands/checkout-editorial-branch.sh` checks out the editorial branch which is [a parameter in the job's trigger.](https://github.com/wordpress-mobile/WordPress-Android/pull/18398/files#diff-e038faf14393bbaff29f4f7842f986b79a4ae5eeafa1c9d0d34d9cef7f86fb08R62)
4. We are temporarily using the `remove/git-push-actions` `release-toolkit` branch [which is already approved](https://github.com/wordpress-mobile/release-toolkit/pull/472). I intend to create the new release tomorrow. As a result of this, you'll notice a few additions of `push_to_git_remote`.
5. The rest of the changes are in the release & localization lanes. They are mostly related to creating an intermediate branch, pushing the changes to remote and creating a pull request.
6. Note that downloading translations will likely not work well in CI in this current iteration. That's because we get `429: Too Many Requests` errors as we try to download several translations and we need to manually confirm to retry. I'm [working on a possible fix for this](https://github.com/wordpress-mobile/release-toolkit/pull/474), but it won't be part of this PR.

**To test:**

Unfortunately there is no way to test the full release flow directly in WPAndroid. However, we used [release-toolkit-demo](https://github.com/wordpress-mobile/release-toolkit-demo) project to test this implementation before bringing it to WPAndroid. So, hopefully any issues we might experience will be minimal.

I was able the test updating release notes flow and it verified that the general setup is working correctly. It's a bit cumbersome to test this and since it's in CI, there isn't anything new to expect, so I suggest skipping this testing step.

We expect some minor issues with the initial implementation, but I am in a position to quickly address them as this project's release manager.

## Regression Notes
1. Potential unintended areas of impact
Parts of release management might break as we move it to CI.

7. What I did to test those areas of impact (or what existing automated tests I relied on)
N/A

8. What automated tests I added (or what prevented me from doing so)
N/A

**PR submission checklist:**

- [x] I have completed the Regression Notes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

**UI Changes testing checklist:**

N/A

### CODE REVIEW

File: .buildkite/commands/configure-git-for-release-management.sh
Comment (oguzkocer) [inline .buildkite/commands/configure-git-for-release-management.sh:1]: We intend to bring this configuration to our agents, so that we don't have to do it in every pipeline.

File: Gemfile
Comment (oguzkocer) [inline Gemfile]: I'll do a proper release before we merge this PR in.

File: fastlane/helpers/github.rb
Comment (oguzkocer) [inline fastlane/helpers/github.rb:1]: I am not sure where the best place for this helper is. I've asked @jkmassel's opinion and he agreed that this approach made sense. However, if you have a suggestion, I'd be happy to apply it.

Note that I initially had this in the `lanes/release_management_in_ci.rb` file but I couldn't get it to work. I am now realizing that if I import that before the other lanes, that approach also would have worked. However, I like the separate file a bit better.

File: fastlane/helpers/github.rb
Comment (oguzkocer) [inline fastlane/helpers/github.rb:3]: [Fastlane uses `ENV["GITHUB_API_TOKEN"]` as the default ](https://docs.fastlane.tools/actions/create_pull_request/)whereas we are using `ENV["GITHUB_TOKEN"]`. Instead of adding another environment variable, I thought it'd be better to default to the existing one.

File: fastlane/helpers/github.rb
Comment (oguzkocer) [inline fastlane/helpers/github.rb:8]: We can add `reviewers` field, but I wanted to skip that for now to not create extra noise. I have tested creating PRs a few times and will likely do a few more. Once everything settles down, it's easy enough to add this field.

File: fastlane/helpers/github.rb
Comment (oguzkocer) [inline fastlane/helpers/github.rb:6]: Fastlane defaults to the current branch. However, it does so by using the [`git_branch`](https://docs.fastlane.tools/actions/git_branch/#git_branch) helper which uses environment variables to override the value. We do not want this behavior which is why we added our own helper in https://github.com/wordpress-mobile/release-toolkit/pull/463.

File: fastlane/lanes/localization.rb
Comment (oguzkocer) [inline fastlane/lanes/localization.rb:106]: The version value is not application specific, in fact it's [fetched](https://github.com/wordpress-mobile/release-toolkit/blob/trunk/lib/fastlane/plugin/wpmreleasetoolkit/helper/android/android_version_helper.rb#L46) from [`version.properties` file](https://github.com/wordpress-mobile/WordPress-Android/blob/trunk/version.properties) which has a shared value. So, there is no harm in moving this outside of the loop and reuse the value in `release_branch` variable.

File: fastlane/lanes/release_management_in_ci.rb
Comment (oguzkocer) [inline fastlane/lanes/release_management_in_ci.rb:1]: Our goal is to remove these triggers from `fastlane`. We'll likely use `curl` for this, so that release managers don't need to have anything setup in their local environment. However, since we already have `buildkite_trigger_build` helper, it made sense to use it during the initial implementation phase.

**Comment by mokagio** [comment]
> Please DO NOT merge this PR, as we need to do a release-toolkit release first.

Created and added the "Do Not Merge" label, and [Peril picked it up](https://github.com/wordpress-mobile/WordPress-Android/pull/18398#issuecomment-1543418006).

image

File: .buildkite/commands/checkout-editorial-branch.sh
Comment (mokagio) [inline .buildkite/commands/checkout-editorial-branch.sh:5]: TIL about this approach to avoid the `set -u` from the shebang to fail the script at this point because of an unbound access. Neat :clap:

File: fastlane/helpers/github.rb
Comment (mokagio) [reply fastlane/helpers/github.rb:3]: Sound 👍

**Review by mokagio — COMMENTED** [review]
I started looking at this but I don't have the energy to give it the attention it deserves right now.

Just a not. Looking at the various new pipelines and how similar they are, it occurred to me that we might make the setup easier to maintain – _as a possible followup!_ – by generating the pipeline. See the [dynamic pipelines docs](https://buildkite.com/docs/pipelines/defining-steps#dynamic-pipelines).

**Comment by oguzkocer** [comment]
> Just a not. Looking at the various new pipelines and how similar they are, it occurred to me that we might make the setup easier to maintain – as a possible followup! – by generating the pipeline. See the [dynamic pipelines docs](https://buildkite.com/docs/pipelines/defining-steps#dynamic-pipelines).

It's definitely worth exploring. I think the main benefit we should look to gain from that should be to get rid of the boilerplate part but still keep the command steps. As long as we keep each file separate - or somehow put it all in a small and easy to read file, then it's a net positive for us. However, if it makes it harder to navigate the code, then I don't think we should do it.

File: fastlane/helpers/github.rb
Comment (jkmassel) [reply fastlane/helpers/github.rb:1]: This seems like a good start – long-term, might be good to see if we can turn it into a release toolkit action – I find that helpers generally end up there!

File: fastlane/helpers/github.rb
Comment (jkmassel) [reply fastlane/helpers/github.rb:3]: This is a good way to go. The reason we use that environment variable is that it's what OctoKit expects FWIW.

File: fastlane/lanes/release_management_in_ci.rb
Comment (jkmassel) [reply fastlane/lanes/release_management_in_ci.rb:1]: I like that this is nicely separated out for easy removal later!

**Review by jkmassel — APPROVED** [review]
Looks solid to me!

File: fastlane/helpers/github.rb
Comment (oguzkocer) [reply fastlane/helpers/github.rb:1]: I agree. To provide some context;

I initially intended to do this and created an internal discussion where Olivier pointed out that there is already a `fastlane` action for creating pull requests. That's when I moved on to adding it to the project instead. I went back and forth on where it belongs but decided to go with this initial implementation because this helper doesn't have much added value. However, at some point, I think it'll make sense to have separate helpers for code freeze, new beta, finalize release etc and then it'll make sense to move it to `release-toolkit` so we can use it in other projects as well.

File: .buildkite/commands/checkout-editorial-branch.sh
Comment (mokagio) [inline .buildkite/commands/checkout-editorial-branch.sh:20]: Is the difference that we wouldn't be able to open the PR if we were on a commit (detached HEAD?)?

File: .buildkite/commands/configure-git-for-release-management.sh
Comment (mokagio) [inline .buildkite/commands/configure-git-for-release-management.sh]: Nitpick, but Buildkite seems a bit generic as the user name in the context of release automation 🤔 

What do you think of 

```suggestion
git config --global user.name "Automattic Release Bot"
```

File: fastlane/helpers/github.rb
Comment (mokagio) [reply fastlane/helpers/github.rb:1]: +1. It makes sense to start local and extract/abstract later.

File: fastlane/lanes/release.rb
Comment (mokagio) [inline fastlane/lanes/release.rb:70]: Nitpick. Since we dropped the `develop`-`trunk` setup, I stopped adding "to trunk" in my PRs because that's the branch where usually release PRs land.

```suggestion
    create_release_management_pull_request('trunk', "Merge #{new_version} code freeze")
```

File: fastlane/lanes/release.rb
Comment (mokagio) [inline fastlane/lanes/release.rb:189]: Same rationale [as above](https://github.com/wordpress-mobile/WordPress-Android/pull/18398/files#r1191844874).

```suggestion
    create_release_management_pull_request('trunk', "Merge #{app_version} final")
```

File: fastlane/lanes/release_management_in_ci.rb
Comment (mokagio) [inline fastlane/lanes/release_management_in_ci.rb]: Nitpick. What do you think of removing "in CI"? After all, "in CI" is the only place where we would see that text.

```suggestion
      message: 'Code Freeze'
```

File: fastlane/lanes/release_management_in_ci.rb
Comment (mokagio) [inline fastlane/lanes/release_management_in_ci.rb]:
```suggestion
      message: 'Complete Code Freeze',
```

Comment (mokagio) [inline fastlane/lanes/release_management_in_ci.rb]:
```suggestion
      message: 'Complete Code Freeze',
```

File: fastlane/lanes/release_management_in_ci.rb
Comment (mokagio) [inline fastlane/lanes/release_management_in_ci.rb]: The previous messages use "Title Case":

```suggestion
      message: 'Finalize Release',
```

File: fastlane/lanes/release_management_in_ci.rb
Comment (mokagio) [inline fastlane/lanes/release_management_in_ci.rb]:
```suggestion
      message: 'New Beta Release',
```

File: fastlane/lanes/release_management_in_ci.rb
Comment (mokagio) [inline fastlane/lanes/release_management_in_ci.rb]:
```suggestion
      message: 'Update Release Notes',
```

**Review by mokagio — APPROVED** [review]
Only left nitpicks that have little to do with the actual implementation. Things are looking pretty good 😄

File: .buildkite/commands/checkout-editorial-branch.sh
Comment (oguzkocer) [reply .buildkite/commands/checkout-editorial-branch.sh:20]: We can't add a new commit in a detached HEAD. When we update the app store strings, we need to create new commits and push it to remote.

File: fastlane/lanes/release.rb
Comment (oguzkocer) [reply fastlane/lanes/release.rb:70]: With release management in CI, not all PRs target the `trunk` branch. Some of them target the release branch, so I think it's useful to have this information.

File: fastlane/lanes/release.rb
Comment (oguzkocer) [reply fastlane/lanes/release.rb:189]: Same rationale [as above](https://github.com/wordpress-mobile/WordPress-Android/pull/18398#discussion_r1191862103). 😅

File: fastlane/lanes/release_management_in_ci.rb
Comment (oguzkocer) [reply fastlane/lanes/release_management_in_ci.rb]: I added this to differentiate some of the information I was figuring out at the moment. So, I am happy to apply this suggestion.

**Comment by oguzkocer** [comment]
Thank you for the review @mokagio! I've applied all your suggestions except for dropping `to trunk` part of the PR titles. I left [a comment about it](https://github.com/wordpress-mobile/WordPress-Android/pull/18398#discussion_r1191862103), but I'll re-iterate it here. With release management in CI, not all PRs will be targeting `trunk`, so I think it's good to be able to see that information at a glance - especially when the GitHub url is unfurled in Slack or p2.

File: .buildkite/commands/configure-git-for-release-management.sh
Comment (AliSoftware) [inline .buildkite/commands/configure-git-for-release-management.sh:10]: 💡  I wonder if it would be better to make this a bit more generic, like:
```bash
REPO_NO_TRAILING=${BUILDKITE_REPO%/}
git remote set-url origin  ${REPO_NO_TRAILING/#https:\/\/github.com\//git@github:}
```
Or alternatively (inspired by [this](https://github.com/Automattic/a8c-ci-toolkit-buildkite-plugin/pull/51/files#diff-d536c5bec9be6d9bc743cc0175b25457a7417a97a3afded630db75505b91882cR83-R84)):
```bash
REPO_SLUG=$(basename $(dirname "$BUILDKITE_REPO"))/$(basename "$BUILDKITE_REPO%.git")
git remote set-url origin git@github.com:${REPO_SLUG}.git
```

Comment (oguzkocer) [reply .buildkite/commands/configure-git-for-release-management.sh:10]: If this was in a toolkit repo, it'd make sense to go with a generic version. However, since this is part of WordPress-Android, I think this just makes it harder to understand the code.

Furthermore, it looks like I forgot to mention that this is a temporary issue specific to WordPress-Android. We should not be using the `https` url to clone in the first place. I just didn't have a chance to figure out why this is the case for WordPress-Android. I think we used to have a specific configuration for WPAndroid, but I couldn't find it yet.

For example, [we don't need this bit in `release-toolkit-demo`.](https://github.com/wordpress-mobile/release-toolkit-demo/blob/trunk/.buildkite/commands/configure-git.sh) I also checked [FluxC](https://buildkite.com/automattic/wordpress-fluxc-android/builds/2488#01881178-a932-4efc-976f-17dec5307aab/135-139) & [WCAndroid](https://buildkite.com/automattic/woocommerce-android/builds/12037#01881100-1de8-4f3a-92e2-dcc8ada0c193) and they both use the `ssh` url.

Comment (oguzkocer) [reply .buildkite/commands/configure-git-for-release-management.sh:10]: Aaah.. and I think I might have found the culprit. All this time I've been checking the Buildkite settings' `Checkout using:` section which says that we are using `ssh` to checkout. However, I just checked the terraform and the `repository` field for WPAndroid is `https://github.com/wordpress-mobile/WordPress-Android/`. I compared this against WPiOS, WCAndroid & FluxC and they all use the `ssh` url. So, I think we'll be able to get rid of this configuration all together.

Having said that, I hope my previous logic also made sense in terms of keeping the specific url.

Comment (AliSoftware) [reply .buildkite/commands/configure-git-for-release-management.sh:10]: > If this was in a toolkit repo, it'd make sense to go with a generic version. However, since this is part of WordPress-Android, I think this just makes it harder to understand the code.

🤦‍♂️ I've reviewed so many PRs in the past few days that I think I lost track of which repo each PR was about, and since I reviewed a PR about `buildkite-ci` repo just before this one, I think I didn't realize I was now reviewing a WPAndroid repo when I switched to this review 🤦‍♂️🤦  Sorry 😅

Comment (oguzkocer) [reply .buildkite/commands/configure-git-for-release-management.sh:10]: No worries at all 🤗 

> I've reviewed so many PRs in the past few days

I noticed ❤️ - thank you for doing that! 🙇‍♂️

**Pull request merged**
